compiler/noirc_evaluator/src/ssa/opt/unrolling.rs

```diff
@@ -54,52 +54,73 @@ impl Ssa {
-            // but only if the setting indicates we have to, otherwise skip it.
-            let orig_func_and_max_incr_pct = max_bytecode_increase_percent
-                .filter(|_| function.runtime().is_brillig())
-                .map(|max_incr_pct| (function.clone(), max_incr_pct));
-
-            // Try to unroll loops first:
-            let (mut has_unrolled, mut unroll_errors) = function.try_unroll_loops();
-
-            // Keep unrolling until no more errors are found
-            while !unroll_errors.is_empty() {
-                let prev_unroll_err_count = unroll_errors.len();
-
-                // Simplify the SSA before retrying
-                simplify_between_unrolls(function);
-
-                // Unroll again
-                let (new_unrolled, new_errors) = function.try_unroll_loops();
-                unroll_errors = new_errors;
-                has_unrolled |= new_unrolled;
-
-                // If we didn't manage to unroll any more loops, exit
-                if unroll_errors.len() >= prev_unroll_err_count {
-                    return Err(unroll_errors.swap_remove(0));
+        for function in self.functions.values_mut() {
+            // We must be able to unroll ACIR loops at this point.
+            function.try_unroll_loops_iteratively(max_bytecode_increase_percent, false)?;
+        }
+        Ok(self)
+    }
+}
+
+impl Function {
+    /// Try to unroll loops in the function.
+    ///
+    /// Returns an `Err` if it cannot be done, for example because the loop bounds
+    /// cannot be determined at compile time. This can happen during pre-processing,
+    /// in which case `restore_on_error` can be used to leave the function untouched.
+    pub(super) fn try_unroll_loops_iteratively(
+        &mut self,
+        max_bytecode_increase_percent: Option<i32>,
+        restore_on_error: bool,
+    ) -> Result<(), RuntimeError> {
+        // Take a snapshot in case we have to restore it.
+        let orig_function = (max_bytecode_increase_percent.is_some()
+            && self.runtime().is_brillig()
+            || restore_on_error)
+            .then(|| self.clone());
+
+        // Try to unroll loops first:
+        let (mut has_unrolled, mut unroll_errors) = self.try_unroll_loops();
+
+        // Keep unrolling until no more errors are found
+        while !unroll_errors.is_empty() {
+            let prev_unroll_err_count = unroll_errors.len();
+
+            // Simplify the SSA before retrying
+            simplify_between_unrolls(self);
+
+            // Unroll again
+            let (new_unrolled, new_errors) = self.try_unroll_loops();
+            unroll_errors = new_errors;
+            has_unrolled |= new_unrolled;
+
+            // If we didn't manage to unroll any more loops, exit
+            if unroll_errors.len() >= prev_unroll_err_count {
+                if restore_on_error {
+                    *self = orig_function.expect("took snapshot to restore");
                 }
```

Comment:
Do we need to restore on errors? reaching an error would result in a partially unrolled function which would still be useful.

Reply:
You may be right; I saw the comment on `unroll_loop_iteration` and that [this loop](https://github.com/noir-lang/noir/blob/0512ee67cc6e4d963c3c8f34c29de40063aae9b7/compiler/noirc_evaluator/src/ssa/opt/unrolling.rs#L407-L419) can fail, and wasn't sure if it can fail after some unrolls have already happened. Also didn't check if there are other kinds of errors that could leave it in invalid state. Would be nice though if we can salvage some of the work!

Reply:
The integration tests seem to pass without it; removed 🤞